samza-azure/src/main/java/org/apache/samza/system/eventhub/producer/EventHubSystemProducer.java

```diff
@@ -0,0 +1,337 @@
+    EventHubClientWrapper ehClient = eventHubClients.get(destination);
+
+    Instant startTime = Instant.now();
+
+    CompletableFuture<Void> sendResult = sendToEventHub(destination, eventData, getEnvelopePartitionId(envelope),
+            ehClient.getEventHubClient());
+
+    Instant endTime = Instant.now();
+    long latencyMs = Duration.between(startTime, endTime).toMillis();
+    sendLatency.get(destination).update(latencyMs);
+    aggSendLatency.update(latencyMs);
+
+    long messageId = ++this.messageId;
+
+    // Rotate the messageIds
+    if (messageId == Long.MAX_VALUE) {
+      this.messageId = 0;
+    }
+
+    pendingFutures.put(messageId, sendResult);
+
+    // Auto remove the future from the list when they are complete.
+    sendResult.handle((aVoid, throwable) -> {
+        long callbackLatencyMs = Duration.between(endTime, Instant.now()).toMillis();
+        sendCallbackLatency.get(destination).update(callbackLatencyMs);
+        aggSendCallbackLatency.update(callbackLatencyMs);
+        if (throwable != null) {
+          sendErrors.get(destination).inc();
+          aggSendErrors.inc();
+          LOG.error("Send message to event hub: {} failed with exception: ", destination, throwable);
+          sendExceptionOnCallback = throwable;
+        }
+        pendingFutures.remove(messageId);
+        return aVoid;
+      });
+  }
+
+  private CompletableFuture<Void> sendToEventHub(String streamName, EventData eventData, Object partitionKey,
+                                                 EventHubClient eventHubClient) {
+    if (partitioningMethod == PartitioningMethod.EVENT_HUB_HASHING) {
+      return eventHubClient.send(eventData, convertPartitionKeyToString(partitionKey));
+    } else if (partitioningMethod == PartitioningMethod.PARTITION_KEY_AS_PARTITION) {
+      if (!(partitionKey instanceof Integer)) {
+        String msg = "Partition key should be of type Integer";
+        LOG.error(msg);
+        throw new SamzaException(msg);
+      }
+
+      PartitionSender sender = getPartitionSender(streamName, (int) partitionKey, eventHubClient);
+      return sender.send(eventData);
+    } else {
+      throw new SamzaException("Unknown partitioning method " + partitioningMethod);
+    }
+  }
+
+  private Object getEnvelopePartitionId(OutgoingMessageEnvelope envelope) {
+    return envelope.getPartitionKey();
+  }
+
+  private String convertPartitionKeyToString(Object partitionKey) {
+    if (partitionKey instanceof String) {
+      return (String) partitionKey;
+    } else if (partitionKey instanceof Integer) {
+      return String.valueOf(partitionKey);
+    } else if (partitionKey instanceof byte[]) {
+      return new String((byte[]) partitionKey, Charset.defaultCharset());
+    } else {
+      throw new SamzaException("Unsupported key type: " + partitionKey.getClass().toString());
+    }
+  }
+
+  private PartitionSender getPartitionSender(String streamName, int partition, EventHubClient eventHubClient) {
```

Comment:
getOrCreate

Reply:
Why is the creation happening during send? Should it happen once during start instead?